readme.md
# Viva Payments for Laravel

[![Latest Version on Packagist](https://img.shields.io/packagist/v/sebdesign/laravel-viva-payments.svg?style=flat-square)](https://packagist.org/packages/sebdesign/laravel-viva-payments)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
[![Build Status](https://img.shields.io/github/workflow/status/sebdesign/laravel-viva-payments/PHP%20Tests?style=flat-square)](https://github.com/sebdesign/laravel-viva-payments/actions)
[![Quality Score](https://img.shields.io/scrutinizer/g/sebdesign/laravel-viva-payments.svg?style=flat-square)](https://scrutinizer-ci.com/g/sebdesign/laravel-viva-payments)
[![Scrutinizer Coverage](https://img.shields.io/scrutinizer/coverage/g/sebdesign/laravel-viva-payments.svg?style=flat-square)](https://scrutinizer-ci.com/g/sebdesign/laravel-viva-payments)

[![VivaPayments logo](https://www.vivawallet.com/assets/vw-logo.svg "Viva Wallet logo")](https://www.vivawallet.com/)

This package provides an interface for the Viva Wallet Payment API. It handles the **Smart Checkout** integration, the **ISV Payment API**, and **Webhooks**.

Check out the official Viva Wallet Developer Portal for detailed instructions on the APIs and more: https://developer.vivawallet.com

**Note:** This project is not a official package, and I'm not affiliated with Viva Payments in any way.

## Setup

#### Installation

Install the package through Composer.

This package requires PHP 8.1 and Laravel 9.0 or higher, and uses Guzzle 7.0 to make API calls. Use the appropriate version according to your dependencies.

| Viva Payments for Laravel   | Guzzle     | Laravel |
|-----------------------------|------------|---------|
| ^6.0                        | ^7.0       | ^9.0    |
| ^5.2                        | ^7.0       | ^9.0    |
| ^5.1                        | ^7.0       | ^8.0    |
| ^5.0                        | ^6.0\|^7.0 | ^7.0    |
| ^4.3                        | ^6.0\|^7.0 | ^7.0    |
| ~4.1                        | ~6.0       | ~7.0    |
| ~4.0                        | ~6.0       | ~6.0    |
| ~3.0                        | ~6.0       | ~5.5    |
| ~2.0                        | ~6.0       | ~5.0    |
| ~1.0                        | ~5.0       | ~5.0    |

```
composer require sebdesign/laravel-viva-payments
```

#### Service Provider

The package will automatically register its service provider.

#### Configuration

Add the following array in your `config/services.php`.

```php
'viva' => [
    'api_key' => env('VIVA_API_KEY'),
    'merchant_id' => env('VIVA_MERCHANT_ID'),
    'environment' => env('VIVA_ENVIRONMENT', 'production'),
    'client_id' => env('VIVA_CLIENT_ID'),
    'client_secret' => env('VIVA_CLIENT_SECRET'),
    'isv_partner_id' => env('VIVA_ISV_PARTNER_ID'),
    'isv_partner_api_key' => env('VIVA_ISV_PARTNER_API_KEY'),
],
```

The `api_key` and `merchant_id` can be found in the *Settings > API Access* section of your profile.

The `client_id` and `client_secret` are needed for the *Smart Checkout*. You can generate the *Smart Checkout Credentials* in the *Settings > API Access* section of your profile.

The `isv_partner_id` and `isv_partner_api_key` are required for using the **ISV Payment API** with Basic authentication.

> Read more about API authentication on the Developer Portal: https://developer.vivawallet.com/getting-started/find-your-account-credentials/client-smart-checkout-credentials/

The `environment` can be either `production` or `demo`.

## Smart Checkout

> Read more about the Smart Checkout process on the Developer portal: https://developer.vivawallet.com/smart-checkout/

The `\Sebdesign\VivaPayments\Facades\Viva` facade provides all the methods needed to interact with the Smart Checkout integration.

The following guide will walk you through the necessary steps:

#### Create the payment order

The amount requested in cents is required. All the other parameters are optional. Check out the [request body schema](https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Payments/paths/~1checkout~1v2~1orders/post).

```php
use Sebdesign\VivaPayments\Facades\Viva;

$orderCode = Viva::orders()->create(
    order: new CreatePaymentOrder(amount: 1000),
);
```

#### Redirect to Smart Checkout

```php
use Sebdesign\VivaPayments\Facades\Viva;

$redirectUrl = Viva::orders()->redirectUrl(
    ref: $orderCode,
    color: '0000ff',
    paymentMethod: 23,
);

return redirect()->away(path: $redirectUrl);
```

#### Verify payment

```php
use Sebdesign\VivaPayments\Facades\Viva;

$response = Viva::transactions()->retrieve(transactionId: request('t'));
```

### Full example

```php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\RedirectResponse;
use Illuminate\Http\Request;
use Sebdesign\VivaPayments\Enums\TransactionStatus;
use Sebdesign\VivaPayments\Facades\Viva;
use Sebdesign\VivaPayments\Requests\CreatePaymentOrder;
use Sebdesign\VivaPayments\Requests\Customer;
use Sebdesign\VivaPayments\VivaException;

class CheckoutController extends Controller
{
    /**
     * Create a payment order and redirect to the checkout page.
     */
    public function checkout(): RedirectResponse
    {
        try {
            $orderCode = Viva::orders()->create(new CreatePaymentOrder(
                amount: 1000,
                customerTrns: 'Short description of purchased items/services to display to your customer',
                customer: new Customer(
                    email: 'johdoe@vivawallet.com',
                    fullName: 'John Doe',
                    countryCode: 'GB',
                    requestLang: 'en-GB',
                ),
            ));
        } catch (VivaException $e) {
            report($e);

            return back()->withErrors($e->getMessage());
        }

        $redirectUrl = Viva::orders()->redirectUrl(
            ref: $orderCode,
            color: '0000ff',
            paymentMethod: 23,
        );

        return redirect()->away($redirectUrl);
    }

    /**
     * Redirect from the checkout page and get the order details from the API.
     */
    public function confirm(Request $request): RedirectResponse
    {
        try {
            $transaction = Viva::transactions()->retrieve($request->input('t'));
        } catch (VivaException $e) {
            report($e);

            return back()->withErrors($e->getMessage());
        }

        $status = match ($transaction->statusId) {
            case TransactionStatus::PaymentPending: 'The order is pending.',
            case TransactionStatus::PaymentSuccessful: 'The order is paid.',
            case TransactionStatus::Error: 'The order was not paid.',
        }

        return view('order/success', compact('status'));
    }
}
```

## Handling Webhooks

Viva Payments supports Webhooks, and this package offers a controller which verifies and handles incoming notification events.

> Read more about the Webhooks on the Developer Portal: https://developer.vivawallet.com/webhooks-for-payments/

### Define the route

In your `routes/web.php` define the following route for each webhook you have in your profile, replacing the URI(s) and your controller(s) accordingly.

```php
Route::get('viva/webhooks', [WebhookController::class, 'verify']);
Route::post('viva/webhooks', [WebhookController::class, 'handle']);
```

### Exclude from CSRF protection

Don't forget to add your webhook URI(s) to the `$except` array on your `VerifyCsrfToken` middleware.

```php
<?php

namespace App\Http\Middleware;

use Illuminate\Foundation\Http\Middleware\VerifyCsrfToken as Middleware;

class VerifyCsrfToken extends Middleware
{
    /**
     * The URIs that should be excluded from CSRF verification.
     *
     * @var array<int, string>
     */
    protected $except = ['viva/webhooks'];
}
```

### Handle webhook events

To handle any request from Viva Wallet, you may listen to the `WebhookEvent`.
According to the `EventTypeId`, you can handle any event.

```php
use Sebdesign\VivaPayments\Enums\WebhookEventType;
use Sebdesign\VivaPayments\Events\WebhookEvent;

class EventServiceProvider
{
    protected $listen = [
        WebhookEvent::class => [WebhookEventListener::class],
    ];
}

class WebhookEventListener
{
    public function handle(WebhookEvent $event): void
    {
        match ($event->EventTypeId) {
            WebhookEventType::TransactionPaymentCreated => ...,
            WebhookEventType::TransactionFailed => ...,
            WebhookEventType::TransactionReversalCreated => ...,
            default => ...,
        };
    }
}
```

The `EventData` property contains an object with the actual notification.
The class of the object depends on the notification type:

| Event                       | Class                       |
|-----------------------------|-----------------------------|
| Transaction Payment Created | `TransactionPaymentCreated` |
| Transaction Failed          | `TransactionFailed`         |
| All other events            | `stdClass`                  |

In addition, the `TransactionPaymentCreated` and `TransactionFailed` events are dispatched. You can listen to these specific events instead of listening to the `WebhookEvent`.

```php
use Sebdesign\VivaPayments\Enums\WebhookEventType;
use Sebdesign\VivaPayments\Events\TransactionFailed;
use Sebdesign\VivaPayments\Events\TransactionPaymentCreated;

class EventServiceProvider
{
    protected $listen = [
        TransactionPaymentCreated::class => [
            ConfirmOrder::class,
        ],
        TransactionFailed::class => [
            CancelOrder::class,
        ],
    ];
}

class ConfirmOrder
{
    public function handle(TransactionPaymentCreated $event): void
    {
        //
    }
}

class CancelOrder
{
    public function handle(TransactionFailed $event): void
    {
        //
    }
}
```

## Payment API reference

All methods accept a `$guzzleOptions` array argument as their last parameter. This argument is entirely optional, and it allows you to specify additional request options to the `Guzzle` client.

### Orders

#### Create a payment order

> See: https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Payments/paths/~1api~1orders~1{orderCode}/get

```php
use Sebdesign\VivaPayments\Facades\Viva;
use Sebdesign\VivaPayments\Requests\CreatePaymentOrder;
use Sebdesign\VivaPayments\Requests\Customer;

$orderCode = Viva::orders()->create(
    order: new CreatePaymentOrder(
        amount: 1000,
        customerTrns: 'Short description of purchased items/services to display to your customer',
        customer: new Customer(
            email: 'johdoe@vivawallet.com',
            fullName: 'John Doe',
            phone: '+30999999999',
            countryCode: 'GB',
            requestLang: 'en-GB',
        ),
        paymentTimeOut: 300,
        preauth: false,
        allowRecurring: false,
        maxInstallments: 12,
        paymentNotification: true,
        tipAmount: 100,
        disableExactAmount: false,
        disableCash: true,
        disableWallet: true,
        sourceCode: '1234',
        merchantTrns: 'Short description of items/services purchased by customer',
        tags: [
            'tags for grouping and filtering the transactions',
            'this tag can be searched on VivaWallet sales dashboard',
            'Sample tag 1',
            'Sample tag 2',
            'Another string',
        ],
        cardTokens: ['ct_5d0a4e3a7e04469f82da228ca98fd661'],
    ),
    guzzleOptions: [],
);
```

##### Get the redirect URL

> See: https://developer.vivawallet.com/smart-checkout/smart-checkout-integration/#step-2-redirect-the-customer-to-smart-checkout-to-pay-the-payment-order

```php
use Sebdesign\VivaPayments\Facades\Viva;

$url = Viva::orders()->redirectUrl(
    ref: $orderCode,
    color: '0000ff',
    paymentMethod: 23,
);
```

### Transactions

#### Retrieve a transaction

> See: https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Transactions/paths/~1checkout~1v2~1transactions~1{transactionId}/get

```php
use Sebdesign\VivaPayments\Facades\Viva;

$transaction = Viva::transactions()->retrieve(
    transactionId: 'c90d4902-6245-449f-b2b0-51d99cd09cfe',
    guzzleOptions: [],
);
```

#### Create a recurring transaction

> See: https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Transactions/paths/~1api~1transactions~1{Id}/post

```php
use Sebdesign\VivaPayments\Facades\Viva;
use Sebdesign\VivaPayments\Requests\CreateRecurringTransaction;

$response = Viva::transactions()->createRecurring(
    transactionId: '252b950e-27f2-4300-ada1-4dedd7c17904',
    transaction: new CreateRecurringTransaction(
        amount: 100,
        installments: 1,
        customerTrns: 'A description of products / services that is displayed to the customer',
        merchantTrns: 'Your merchant reference',
        sourceCode: '6054',
        tipAmount: 0,
    ),
    guzzleOptions: [],
);
```

### OAuth

#### Request access token

> See: https://developer.vivawallet.com/authentication/#step-2-request-access-token

You don't need to call this method, because the client requests the access token automatically when needed.
However, you can specify the client credentials at runtime if you want.

```php
use Sebdesign\VivaPayments\Facades\Viva;

Viva::withOAuthCredentials(
    clientId: 'client_id',
    clientSecret: 'client_secret',
);
```

If you need to request access tokens manually, you can use the `requestToken` method.
This method returns the token as an `AccessToken` object.

```php
use Sebdesign\VivaPayments\Facades\Viva;

// Using `client_id` and `client_secret` from `config/services.php`:
$token = Viva::oauth()->requestToken();

// Using custom client credentials
$token = Viva::oauth()->requestToken(
    clientId: 'client_id',
    clientSecret: 'client_secret',
    guzzleOptions: [],
);
```

### Use an existing access token

If you are storing the token somewhere, e.g. in your database or in the cache, you can set the access token string on the client to be used as a Bearer token.

```php
use Sebdesign\VivaPayments\Facades\Viva;

Viva::withToken(token: 'eyJhbGciOiJSUzI1...');
```

### Cards

#### Create card token

> See: https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Transactions/paths/~1acquiring~1v1~1cards~1tokens/post

```php
use Sebdesign\VivaPayments\Facades\Viva;

$cardToken = Viva::cards()->createToken(
    transactionId: '6cffe5bf-909c-4d69-b6dc-2bef1a6202f7',
    guzzleOptions: [],
);
```

### Webhooks

##### Get an authorization code

> See: https://developer.vivawallet.com/webhooks-for-payments/#generate-a-webhook-verification-key

```php
use Sebdesign\VivaPayments\Facades\Viva;

$key = Viva::webhooks()->getVerificationKey(
    guzzleOptions: [],
);
```

## ISV Payment API Reference

The ISV Payment API methods are available through the `Viva::isv()` service.

### Orders

#### Create a payment order

> See: https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Payments/paths/~1checkout~1v2~1isv~1orders/post

```php
use Sebdesign\VivaPayments\Facades\Viva;
use Sebdesign\VivaPayments\Requests\CreatePaymentOrder;
use Sebdesign\VivaPayments\Requests\Customer;

$orderCode = Viva::isv()->orders()->create(
    order: new CreatePaymentOrder(
        amount: 1000,
        customerTrns: 'Short description of purchased items/services to display to your customer',
        customer: new Customer(
            email: 'johdoe@vivawallet.com',
            fullName: 'John Doe',
            phone: '+30999999999',
            countryCode: 'GB',
            requestLang: 'en-GB',
        ),
        paymentTimeOut: 300,
        preauth: false,
        allowRecurring: false,
        maxInstallments: 12,
        paymentNotification: true,
        tipAmount: 100,
        disableExactAmount: false,
        disableCash: true,
        disableWallet: true,
        sourceCode: '1234',
        merchantTrns: 'Short description of items/services purchased by customer',
        tags: [
            'tags for grouping and filtering the transactions',
            'this tag can be searched on VivaWallet sales dashboard',
            'Sample tag 1',
            'Sample tag 2',
            'Another string',
        ],
        isvAmount: 10,
        resellerSourceCode: '2345',
    ),
    guzzleOptions: [],
);
```

### Transactions

#### Retrieve a transaction

> See: https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Retrieve-Transactions/paths/~1checkout~1v2~1isv~1transactions~1{transactionId}?merchantId={merchantId}/get

```php
use Sebdesign\VivaPayments\Facades\Viva;

$transaction = Viva::isv()->transactions()->retrieve(
    transactionId: 'c90d4902-6245-449f-b2b0-51d99cd09cfe',
    guzzleOptions: [],
);
```

#### Create a recurring transaction

> See: https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Recurring-Payments/paths/~1api~1transactions~1{id}/post

```php
use Sebdesign\VivaPayments\Facades\Viva;
use Sebdesign\VivaPayments\Requests\CreateRecurringTransaction;

Viva::withBasicAuthCredentials(
    config('services.viva.isv_partner_id').':'.config('services.viva.merchant_id'),
    config('services.viva.isv_partner_api_key'),
);

$transaction = Viva::isv()->transactions()->createRecurring(
    transactionId: 'c90d4902-6245-449f-b2b0-51d99cd09cfe',
    transaction: new CreateRecurringTransaction(
        amount: 100,
        isvAmount: 1,
        customerTrns: 'A description of products / services that is displayed to the customer',
        merchantTrns: 'Your merchant reference',
        sourceCode: '4929333',
        resellerSourceCode: '1565',
    ),
    guzzleOptions: [],
);
```

## Exceptions

When the VivaPayments API returns an error, a `Sebdesign\VivaPayments\VivaException` is thrown.

For any other HTTP error a `GuzzleHttp\Exception\ClientException` is thrown.

## Tests

Unit tests are triggered by running `phpunit --group unit`.

To run functional tests you have to include a `.env` file in the root folder, containing the credentials (`VIVA_API_KEY`, `VIVA_MERCHANT_ID`, `VIVA_CLIENT_ID`, `VIVA_CLIENT_SECRET`), in order to hit the VivaPayments demo API. Then run `phpunit --group functional` to trigger the tests.


CHANGELOG.md
# Changelog

All notable changes to `sebdesign/laravel-viva-payments` will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [6.0.0-alpha.6] - 2022-12-30

### Fixed

-   Make `cardUniqueReference` nullable on transaction response ([#29](https://github.com/sebdesign/laravel-viva-payments/issues/29))

## [6.0.0-alpha.5] - 2022-11-30

### Fixed

-   Add `TransactionTypeId` to recurring transaction response

## [6.0.0-alpha.4] - 2022-11-29

### Fixed

-   Make `merchantTrns` nullable on transaction response ([#28](https://github.com/sebdesign/laravel-viva-payments/issues/28))

## [6.0.0-alpha.3] - 2022-11-18

### Added

-   Improve exceptions
-   Create recurring transactions for ISV

## [6.0.0-alpha.2] - 2022-11-10

-   Extract service objects in subdirectory
-   Rename `SmartCheckout` facade to `Viva`
-   Create payment orders for ISV
-   Retrieve transaction by id for ISV

## [6.0.0-alpha.1] - 2022-11-08

### Added

-   Implement Smart Checkout
-   Create card tokens
-   Add support for PHP 8.2

### Changes

-   Dispatch events instead of extending controller for handling webhooks

### Removed

-   Remove Simple Checkout
-   Remove Native Checkout
-   Remove Redirect Checkout
-   Remove deprecated APIs
-   Drop support for PHP 8.0 and below
-   Drop support for Laravel 8 and below

## [5.2.0] - 2022-01-13

-   Add support for Laravel 9

## [5.1.6] - 2021-11-27

-   Add support for PHP 8.0 and 8.1

## [5.1.5] - 2021-10-17

-   Fix relative URI exception from guzzle

## [5.1.4] - 2021-05-28

-   Use POST instead of GET method in capturePreAuthTransaction API [#24](https://github.com/sebdesign/laravel-viva-payments/pull/24)

## [5.1.3] - 2021-02-19

-   Update to TLS v1.2

## [5.1.2] - 2020-12-09

-   Switch to GitHub Actions

## [5.1.1] - 2020-09-29

-   Fix NativeCheckout::chargeTokenUsingCardToken [#18](https://github.com/sebdesign/laravel-viva-payments/issues/18)

## [5.1.0] - 2020-09-07

-   Add support for Laravel 8

## [5.0.0] - 2020-09-01

-   Add OAuth authentication
-   Add Simple Checkout
-   Add Native Checkout v2
-   Remove Native Checkout v1
-   Remove Mobile Checkout

## [4.4.0] - 2020-08-29

-   Allow additional options for the Guzzle client

## [4.3.1] - 2020-07-15

-   Remove extraneous single quote ([#17](https://github.com/sebdesign/laravel-viva-payments/pull/17)) (thanks [@adrianblynch](https://github.com/adrianblynch))

## [4.3.0] - 2020-07-08

-   Allow Guzzle 7

## [4.2.0] - 2020-05-05

-   Allow recurring payments when creating orders
-   Fix documentation links

## [4.1.1] - 2020-03-21

-   Change `orderCode` parameter to `s` after redirect in documentation

## [4.1.0] - 2020-03-03

-   Add support for Laravel 7

## [4.0.0] - 2019-09-02

-   Add support for Laravel 6.0
-   Drop support for Laravel <5.5 and PHP <7.1

## [3.2.0] - 2019-02-15

-   Add support for Laravel 5.8

## [3.1.0] - 2018-07-31

-   Add support for Laravel 5.7

## [3.0.1] - 2018-06-01

-   Use TLSv1 cipher list if cURL doesn't use NSS

## [3.0.0] - 2017-09-14

-   Add support for Laravel 5.5

## [2.0.1] - 2016-11-02

-   Use https everywhere

## [2.0.0] - 2016-04-26

-   Use Guzzle 6

## [1.0.0] - 2016-04-29

-   Initial release


src/Transaction.php
<?php

namespace Sebdesign\VivaPayments;

use GuzzleHttp\RequestOptions;

class Transaction
{
    /**
     * Transaction types.
     */

    // A Capture event of a preAuthorized transaction
    const CAPTURE_FROM_PREAUTH = 0;

    // Authorization hold
    const PREAUTH = 1;

    // Refund transaction
    const REFUND_CARD = 4;

    // Card payment transaction
    const CHARGE_CARD = 5;

    // A card payment that will be done with installments
    const CHARGE_CARD_WITH_INSTALLMENTS = 6;

    // A payment cancelation
    const VOID = 7;

    // A Wallet Payment
    const WALLET_CHARGE = 9;

    // A Refund of a Wallet Payment
    const WALLET_REFUND = 11;

    // Refund transaction for a claimed transaction
    const CLAIM_REFUND = 13;

    // Payment made through the DIAS system
    const DIAS_PAYMENT = 15;

    // Cash Payments, through the Viva Payments Authorised Resellers Network
    const CASH_PAYMENT = 16;

    // A Refunded installment
    const REFUND_INSTALLMENTS = 18;

    // Clearance of a transactions batch
    const CLEARANCE = 19;

    // Bank Transfer command from the merchant's wallet to their IBAN
    const BANK_TRANSFER = 24;

    /**
     * Transaction statuses.
     */

    // The transaction was not completed because of an error
    const ERROR = 'E';

    // The transaction is in progress
    const PROGRESS = 'A';

    // The cardholder has disputed the transaction with the issuing Bank
    const DISPUTED = 'M';

    // Dispute Awaiting Response
    const DISPUTE_AWAITING = 'MA';

    // Dispute in Progress
    const DISPUTE_IN_PROGRESS = 'MI';

    // A disputed transaction has been refunded (Dispute Lost)
    const DISPUTE_REFUNDED = 'ML';

    // Dispute Won
    const DISPUTE_WON = 'MW';

    // Suspected Dispute
    const DISPUTE_SUSPECTED = 'MS';

    // The transaction was cancelled by the merchant
    const CANCELED = 'X';

    // The transaction has been fully or partially refunded
    const REFUNDED = 'R';

    // The transaction has been completed successfully
    const COMPLETED = 'F';

    /**
     * @var \Sebdesign\VivaPayments\Client
     */
    protected $client;

    /**
     * Constructor.
     */
    public function __construct(Client $client)
    {
        $this->client = $client;
    }

    /**
     * Create a new transaction.
     *
     * @see https://developer.vivawallet.com/online-checkouts/simple-checkout/#step-2-make-the-charge
     *
     * @param  array  $parameters
     * @param  array  $guzzleOptions  Additional parameters for the Guzzle client
     * @return \stdClass
     */
    public function create(array $parameters, array $guzzleOptions = [])
    {
        return $this->client->post(
            $this->client->getUrl()->withPath('/api/transactions'),
            array_merge_recursive(
                [RequestOptions::JSON => $parameters],
                $this->client->authenticateWithBasicAuth(),
                $guzzleOptions
            )
        );
    }

    /**
     * Create a recurring transaction.
     *
     * This API call allows you to make a new payment by either committing
     * an already authorized transaction or by making a recurring payment.
     * The latter is only permitted if the following two conditions are met:
     *
     * - The cardholder has already been charged successfully in the past
     * - The cardholder has agreed to allow recurring payments on their card
     *
     * @see https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Transactions/paths/~1api~1transactions~1{Id}/post
     *
     * @param  string  $id  The transaction's unique ID
     * @param  int  $amount  The amount requested in the currency's smallest unit of measurement
     * @param  array  $parameters  Transaction parameters
     * @param  array  $guzzleOptions  Additional parameters for the Guzzle client
     * @return \stdClass
     */
    public function createRecurring(
        string $id,
        int $amount,
        array $parameters = [],
        array $guzzleOptions = []
    ) {
        $parameters = array_merge(['amount' => $amount], $parameters);

        return $this->client->post(
            $this->client->getUrl()->withPath("/api/transactions/{$id}"),
            array_merge_recursive(
                [RequestOptions::JSON => $parameters],
                $this->client->authenticateWithBasicAuth(),
                $guzzleOptions
            )
        );
    }

    /**
     * Get details for a single transaction, identified by its transactionId.
     *
     * @see https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Transactions/paths/~1api~1transactions~1{Id}/get
     *
     * @param  string  $id
     * @param  array  $guzzleOptions  Additional parameters for the Guzzle client
     * @return array
     */
    public function get(string $id, array $guzzleOptions = []): array
    {
        $response = $this->client->get(
            $this->client->getUrl()->withPath("/api/transactions/{$id}"),
            array_merge_recursive(
                $this->client->authenticateWithBasicAuth(),
                $guzzleOptions
            )
        );

        return $response->Transactions;
    }

    /**
     * Get details for all transactions for a given payment order.
     *
     * @see https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Transactions/paths/~1api~1transactions~1{Id}/get
     *
     * @param  int  $ordercode
     * @param  array  $guzzleOptions  Additional parameters for the Guzzle client
     * @return array
     */
    public function getByOrder($ordercode, array $guzzleOptions = []): array
    {
        $parameters = ['ordercode' => $ordercode];

        $response = $this->client->get(
            $this->client->getUrl()->withPath('/api/transactions'),
            array_merge_recursive(
                [RequestOptions::QUERY => $parameters],
                $this->client->authenticateWithBasicAuth(),
                $guzzleOptions
            )
        );

        return $response->Transactions;
    }

    /**
     * List of all transactions that occurred on a given date.
     *
     * @see https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Transactions/paths/~1api~1transactions~1{Id}/get
     *
     * @param  \DateTimeInterface|string  $date
     * @param  array  $guzzleOptions  Additional parameters for the Guzzle client
     * @return array
     */
    public function getByDate($date, array $guzzleOptions = []): array
    {
        $parameters = ['date' => $this->formatDate($date)];

        $response = $this->client->get(
            $this->client->getUrl()->withPath('/api/transactions'),
            array_merge_recursive(
                [RequestOptions::QUERY => $parameters],
                $this->client->authenticateWithBasicAuth(),
                $guzzleOptions
            )
        );

        return $response->Transactions;
    }

    /**
     * List of all transactions that occurred on a specific clearance date.
     *
     * @see https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Transactions/paths/~1api~1transactions~1{Id}/get
     *
     * @param  \DateTimeInterface|string  $clearancedate
     * @param  array  $guzzleOptions  Additional parameters for the Guzzle client
     * @return array
     */
    public function getByClearanceDate($clearancedate, array $guzzleOptions = []): array
    {
        $parameters = ['clearancedate' => $this->formatDate($clearancedate)];

        $response = $this->client->get(
            $this->client->getUrl()->withPath('/api/transactions'),
            array_merge_recursive(
                [RequestOptions::QUERY => $parameters],
                $this->client->authenticateWithBasicAuth(),
                $guzzleOptions
            )
        );

        return $response->Transactions;
    }

    /**
     * List of all transactions for a given Source Code for a specific date.
     *
     * @see https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Transactions/paths/~1api~1transactions~1{Id}/get
     *
     * @param  string  $sourcecode
     * @param  \DateTimeInterface|string  $date
     * @param  array  $guzzleOptions  Additional parameters for the Guzzle client
     * @return array
     */
    public function getBySourceCode($sourcecode, $date, array $guzzleOptions = []): array
    {
        $parameters = [
            'sourcecode' => $sourcecode,
            'date' => $this->formatDate($date),
        ];

        $response = $this->client->get(
            $this->client->getUrl()->withPath('/api/transactions'),
            array_merge_recursive(
                [RequestOptions::QUERY => $parameters],
                $this->client->authenticateWithBasicAuth(),
                $guzzleOptions
            )
        );

        return $response->Transactions;
    }

    /**
     * Cancel or refund a payment.
     *
     * @see https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Transactions/paths/~1api~1transactions~1{Id}/delete
     *
     * @param  string  $id  The transaction's unique ID
     * @param  int  $amount  The amount that will be refunded in the currency's smallest denomination (e.g amount in pounds x 100)
     * @param  string|null  $sourceCode  The source from which the funds will be withdrawn. Each source is linked to a wallet. If no sourceCode is set then the funds will be withdrawn from the primary wallet.
     * @param  array  $guzzleOptions  Additional parameters for the Guzzle client
     * @return \stdClass
     */
    public function cancel(
        string $id,
        int $amount,
        $sourceCode = null,
        array $guzzleOptions = []
    ) {
        $parameters = array_merge(
            ['amount' => $amount],
            $sourceCode ? ['sourceCode' => $sourceCode] : []
        );

        return $this->client->delete(
            $this->client->getUrl()->withPath("/api/transactions/{$id}"),
            array_merge_recursive(
                [RequestOptions::QUERY => $parameters],
                $this->client->authenticateWithBasicAuth(),
                $guzzleOptions
            )
        );
    }

    /**
     * Format a date object to string.
     *
     * @param  \DateTimeInterface|string  $date
     * @return string
     */
    protected function formatDate($date): string
    {
        if ($date instanceof \DateTimeInterface) {
            return $date->format('Y-m-d');
        }

        return $date;
    }
}
